### PR TITLE
#511 corrected extension template

### PR DESCRIPTION
**Pull request by dmcassel**
Response codes should be set in $context, not with xdmp:set-response-code().
http://docs.marklogic.com/guide/rest-dev/extensions#id_31884 #511

### CODE REVIEW

**Comment by grtjn** [comment]
Maybe also give an example of how to return an error status? I thought you cannot use output-status for that: http://docs.marklogic.com/guide/rest-dev/extensions#id_33892

**Comment by dmcassel** [comment]
Right, output-status is not for errors. What do you think about a comment up top providing an example and referring to the appropriate documentation?

**Comment by dmcassel** [comment]
Updated as discussed.

File: deploy/sample/rest-ext.sample.xqy [hunk 1]
```diff
@@ -5,11 +5,19 @@ module namespace yourNSAlias = "http://marklogic.com/rest-api/resource/extension
...
-(: 
- : To add parameters to the functions, specify them in the params annotations. 
+(:
+ : To add parameters to the functions, specify them in the params annotations.
  : Example
  :   declare %roxy:params("uri=xs:string", "priority=xs:int") yourNSAlias:get(...)
  : This means that the get function will take two parameters, a string and an int.
+ :
+ : To report errors in your extension, use fn:error(). For details, see
+ : http://docs.marklogic.com/6.0/guide/rest-dev/transforms#id_50497, but here's
```
Comment (grtjn) [inline deploy/sample/rest-ext.sample.xqy]: Interesting. You are pointing to 6.0 docs, and to a section related to REST transforms? :)

Comment (grtjn) [inline deploy/sample/rest-ext.sample.xqy]: This seems the link to the current REST extension equivalent. Looks slightly different? http://docs.marklogic.com/guide/rest-dev/extensions#id_33892

**Comment by heelix** [comment]
For what it is worth, if you specify a "output-status", and return the fn:error in whatever lib is getting invoked, the error will override whatever was specified in the REST endpoint (/rest-api/ext) code.  If one specified a default status code other than the 200/OK, it would not stick.

(Ended up deleting the rest endpoints we did not want to support, which also kicks a good enough error message to the REST consumer.)

**Comment by grtjn** [comment]
That is what I would have expected indeed. The fn:error is pretty strong, it is like throwing an Exception in Java, and the function will not have the chance to return the result.

Not implementing particular methods returns a 405, method not allowed?

**Comment by heelix** [comment]
400 Bad Request

{
errorResponse: {
statusCode: 400
status: "Bad Request"
messageCode: "XDMP-UNDFUN"
message: "XDMP-UNDFUN: (err:XPST0017) xdmp:function(fn:QName("http://marklogic.com/rest-api/resource/xxx","put"), "/marklogic.rest.resource/deliver/assets/xxx.xqy")($context, $service-params, $input) -- Undefined function put()"
}-
}

**Comment by grtjn** [comment]
Sounds like an RFE against MarkLogic.. :)

File: deploy/sample/rest-ext.sample.xqy (hunk 1 above)
Comment (dmcassel) [inline deploy/sample/rest-ext.sample.xqy]: Doh. I was checking how far back some of the stuff goes, got the wrong link. Will fix.

File: deploy/sample/rest-ext.sample.xqy [hunk 2]
```diff
@@ -5,11 +5,20 @@ module namespace yourNSAlias = "http://marklogic.com/rest-api/resource/extension
...
-(: 
- : To add parameters to the functions, specify them in the params annotations. 
+(:
+ : To add parameters to the functions, specify them in the params annotations.
  : Example
  :   declare %roxy:params("uri=xs:string", "priority=xs:int") yourNSAlias:get(...)
  : This means that the get function will take two parameters, a string and an int.
+ :
+ : To report errors in your extension, use fn:error(). For details, see
+ : http://docs.marklogic.com/guide/rest-dev/transforms#id_50497, but here's
```
Comment (grtjn) [inline deploy/sample/rest-ext.sample.xqy]: Still pointing to transforms section? Perhaps use this: http://docs.marklogic.com/guide/rest-dev/extensions#id_33892

**Comment by dmcassel** [comment]
@grtjn got the right link this time

**Comment by grtjn** [comment]
:)

**Pull request merged**